File: ts-src/history/reader.ts
import { existsSync, readFileSync } from "node:fs";
import { getHistoryPath } from "../config/paths.ts";
import type { HistoryEntry } from "./writer.ts";

export interface LoadHistoryOptions {
	limit?: number;
	search?: string;
}

export async function loadHistory(options?: LoadHistoryOptions): Promise<HistoryEntry[]> {
	const path = getHistoryPath();
	if (!existsSync(path)) return [];

	const content = readFileSync(path, "utf-8");
	const lines = content.trim().split("\n").filter(Boolean);

	let entries: HistoryEntry[] = [];
	for (const line of lines) {
		try {
			entries.push(JSON.parse(line));
		} catch {
			// skip malformed lines
		}
	}

	if (options?.search) {
		const term = options.search.toLowerCase();
		entries = entries.filter((e) => e.message_preview.toLowerCase().includes(term));
	}

	if (options?.limit) {
		entries = entries.slice(-options.limit);
	}

	return entries;
}

File: src/history/reader.rs
//! Read filtered history entries.

use super::writer::HistoryEntry;
use crate::config::paths::get_history_path;

#[derive(Debug, Clone, Default)]
pub struct LoadHistoryOptions {
    pub limit: Option<usize>,
    pub search: Option<String>,
}

pub fn load_history(options: &LoadHistoryOptions) -> Vec<HistoryEntry> {
    let path = get_history_path();
    let content = match std::fs::read_to_string(&path) {
        Ok(c) => c,
        Err(_) => return Vec::new(),
    };

    let mut entries: Vec<HistoryEntry> = content
        .lines()
        .filter(|l| !l.trim().is_empty())
        .filter_map(|l| serde_json::from_str(l).ok())
        .collect();

    if let Some(search) = &options.search {
        let term = search.to_lowercase();
        entries.retain(|e| e.message_preview.to_lowercase().contains(&term));
    }
    if let Some(limit) = options.limit {
        if entries.len() > limit {
            entries.drain(..entries.len() - limit);
        }
    }
    entries
}
